certora/specs/SocialRecoveryModule.spec

```
@@ -1,26 +1,208 @@
+rule addGuardianRevertPossibilities(env e, address guardian, uint256 threshold) {
+    bool isGuardian = currentContract.isGuardian(safeContract, guardian);
+
+    currentContract.addGuardianWithThreshold@withrevert(e, safeContract, guardian, threshold);
+    bool isReverted = lastReverted;
+
+    assert isReverted =>
+        isGuardian ||
+        e.msg.sender != safeContract ||
+        e.msg.value != 0 ||
+        guardian == 0 ||
+        guardian == SENTINEL() ||
+        guardian == safeContract ||
+        safeContract.isOwner(guardian) ||
+        threshold == 0 ||
+        threshold >= guardianStorageContract.entries[safeContract].count ||
+        guardianStorageContract.entries[safeContract].count == max_uint256 ||
+        !safeContract.isModuleEnabled(currentContract);
+}
+
+// This integrity rule verifies that if the revokeGuardianWithThreshold(...) executes, then ensure that:
+// - the Social Recovery Module is enabled
+// - the caller to the Module has to be the Safe Contract
+// - the guardian is revoked from the guardian list
+// - the linked list integrity remains,
+// - and no other account (guardian or not) is affected.
+rule revokeGuardiansWorksAsExpected(env e, address guardian, address prevGuardian, uint256 threshold, address otherAccount) {
+    address nextGuardian = guardianStorageContract.entries[safeContract].guardians[guardian];
+
+    bool otherAccountIsGuardian = currentContract.isGuardian(safeContract, otherAccount);
+
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;
+    require currentGuardiansCount > 0;
+
+    requireGuardiansLinkedListIntegrity(guardian);
 
+    currentContract.revokeGuardianWithThreshold(e, safeContract, prevGuardian, guardian, threshold);
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert !currentContract.isGuardian(safeContract, guardian);
+    assert guardianStorageContract.entries[safeContract].guardians[prevGuardian] == nextGuardian;
+    assert guardian != otherAccount => otherAccountIsGuardian == currentContract.isGuardian(safeContract, otherAccount);
+    assert currentGuardiansCount - 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+}
+
+// This integrity rule verifies that the guardian can always be revoked considering ideal conditions.
+rule guardianCanAlwaysBeRevoked(env e, address guardian, address prevGuardian, uint256 threshold) {
     requireSocialRecoveryModuleEnabled();
 
-    safeContract.disableModule@withrevert(e, prevModule, currentContract);
+    // No value should be sent with the transaction.
+    require e.msg.value == 0;
+    // Guardian should not be zero address.
+    require guardian != 0;
+    // New threshold should be greater than 0 and the guardian count should be greater than the threshold.
+    require threshold > 0 && guardianStorageContract.entries[safeContract].count > threshold;
```

Suggestion: IMO what this check is that:
```suggestion
    // New threshold should be greater than 0 and less than or equal to the guardian count after removing.
    require threshold > 0 && threshold <= guardianStorageContract.entries[safeContract].count - 1;
```

---

Actually, I think this is wrong as well :see_no_evil: - you can remove the last guardian and end up with an empty list, so you can't have `threshold > 0` as a requirement. When you remove the last guardian, you end up with `count = threshold = 0`.

---

Made a conditional check based on threshold value 👍🏾